File: src/test/java/io/github/saneea/dvh/utils/ByteSequenceRecognizerTest.java
package io.github.saneea.dvh.utils;

import static org.junit.Assert.assertEquals;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Stream;

import org.junit.Assert;
import org.junit.Test;

import io.github.saneea.dvh.utils.ByteSequenceRecognizer.ByteNode;

public class ByteSequenceRecognizerTest {

	private static final Charset TEST_CHARSET = StandardCharsets.US_ASCII;

	@Test
	public void testTwoBranches() throws IOException {
		String[] sequences = { "123", "124" };
		test("123", Optional.of(0), sequences);
		test("124", Optional.of(1), sequences);

		test("12345678", Optional.of(0), sequences);
		test("12456789", Optional.of(1), sequences);

		test("12", Optional.empty(), sequences);

		test("12X", Optional.empty(), sequences);
		test("12X", Optional.empty(), sequences);

		test("", Optional.empty(), sequences);
	}

	@Test
	public void testDifferentSize() throws IOException {
		String[] sequences = { "123", "1234" };
		test("123", Optional.of(0), sequences);
		test("1234", Optional.of(0), sequences);
	}

	@Test
	public void testEmpty() throws IOException {
		String[] sequences = { "", "123" };
		test("123", Optional.of(0), sequences);
		test("1234", Optional.of(0), sequences);
		test("", Optional.of(0), sequences);
	}

	@Test(expected = IllegalArgumentException.class)
	public void testNoSequences() throws IOException {
		String[] sequences = {};
		test("123", Optional.empty(), sequences);
		test("1234", Optional.empty(), sequences);
		test("", Optional.empty(), sequences);
	}

	@Test
	public void testTwoBranchesLong() throws IOException {
		String[] sequences = { "123abcdef", "124abcdef" };
		test("123abcdef", Optional.of(0), sequences);
		test("124abcdef", Optional.of(1), sequences);

		test("123abcdefXYZ", Optional.of(0), sequences);
		test("124abcdefXYZ", Optional.of(1), sequences);

		test("123abcde", Optional.empty(), sequences);
		test("124abcde", Optional.empty(), sequences);

		test("123abcdeX", Optional.empty(), sequences);
		test("124abcdeX", Optional.empty(), sequences);

		test("", Optional.empty(), sequences);
	}

	@Test
	public void testDepth() throws IOException {
		testDepthInternal();
		testDepthInternal("");
		testDepthInternal("", "1");
		testDepthInternal("1", "");
		testDepthInternal("1");
		testDepthInternal("1", "2");
		testDepthInternal("1", "22");
		testDepthInternal("11", "2");
		testDepthInternal("11", "22");
		testDepthInternal("11", "222");
		testDepthInternal("111", "222");
		testDepthInternal("111", "222", "1111111");
	}

	private void testDepthInternal(String... sequences) throws IOException {
		ByteNode<Integer> node = ByteSequenceRecognizer.ByteNode.fromMap(//
				sequencesMap(//
						sequences));

		int expectedDepth = Stream//
				.of(sequences)//
				.mapToInt(String::length)//
				.max().orElse(0);

		assertEquals(expectedDepth, node.getDepth());
	}

	@Test
	public void testAllBytesValues() throws IOException {

		int branchId = 1;

		for (byte branchByte : allBytes()) {

			int[] sequence = { branchByte & 0xff };
			ByteNode<Integer> sequencesTree = ByteSequenceRecognizer.ByteNode.fromMap(Map.of(sequence, branchId));

			for (byte inputByte : allBytes()) {

				byte[] inputData = { inputByte };
				try (ByteSequenceRecognizer<Integer> bsr = new ByteSequenceRecognizer<>(//
						new ByteArrayInputStream(inputData), //
						sequencesTree)) {

					byte[] actualBytes = bsr.stream().readAllBytes();
					Assert.assertArrayEquals(inputData, actualBytes);

					Optional<Integer> actualResult = bsr.result();
					Optional<Integer> expectedResult = branchByte == inputByte//
							? Optional.of(branchId)//
							: Optional.empty();
					assertEquals(expectedResult, actualResult);
				}
			}
		}

		// Assert.fail("not implemented");
	}

	private static Iterable<Byte> allBytes() {
		return () -> new Iterator<Byte>() {

			Byte next = Byte.MIN_VALUE;

			@Override
			public boolean hasNext() {
				return next != null;
			}

			@Override
			public Byte next() {
				byte ret = next;
				if (next == Byte.MAX_VALUE) {
					next = null;
				} else {
					next++;
				}
				return ret;
			}

		};
	}

	private void test(String inputStr, Optional<Integer> expectedResult, String... sequences) throws IOException {
		try (ByteSequenceRecognizer<Integer> bsr = recognizer(//
				inputStr, //
				ByteSequenceRecognizer.ByteNode.fromMap(//
						sequencesMap(//
								sequences)))) {

			String actualText = asString(bsr.stream().readAllBytes());
			assertEquals(inputStr, actualText);

			Optional<Integer> actualResult = bsr.result();
			assertEquals(expectedResult, actualResult);
		}
	}

	private static String asString(byte[] bytes) {
		return new String(bytes, TEST_CHARSET);
	}

	private static Map<int[], Integer> sequencesMap(String... sequences) {
		Map<int[], Integer> r = new HashMap<>();
		for (int i = 0; i < sequences.length; ++i) {
			r.put(intArray(sequences[i]), i);
		}
		return r;
	}

	private static int[] intArray(String s) {
		return intArray(bytes(s));
	}

	private static int[] intArray(byte[] byteArray) {
		int r[] = new int[byteArray.length];
		for (int i = 0; i < byteArray.length; ++i) {
			r[i] = byteArray[i];
		}
		return r;
	}

	private static ByteSequenceRecognizer<Integer> recognizer(//
			String s, ByteNode<Integer> sequencesTree) throws IOException {
		return new ByteSequenceRecognizer<>(stream(s), sequencesTree);
	}

	private static InputStream stream(String s) {
		return new ByteArrayInputStream(bytes(s));
	}

	private static byte[] bytes(String s) {
		return s.getBytes(TEST_CHARSET);
	}
}

File: src/test/java/io/github/saneea/dvh/utils/ByteSequenceRecognizerTest.kt
package io.github.saneea.dvh.utils

import io.github.saneea.dvh.utils.ByteSequenceRecognizer.ByteNode
import org.junit.Assert
import org.junit.Test
import java.io.ByteArrayInputStream
import java.nio.charset.StandardCharsets

private val TEST_CHARSET = StandardCharsets.US_ASCII

class ByteSequenceRecognizerTest {

    @Test
    fun testTwoBranches() {
        val sequences = listOf("123", "124")
        test("123", 0, sequences)
        test("124", 1, sequences)
        test("12345678", 0, sequences)
        test("12456789", 1, sequences)
        test("12", null, sequences)
        test("12X", null, sequences)
        test("12X", null, sequences)
        test("", null, sequences)
    }

    @Test
    fun testDifferentSize() {
        val sequences = listOf("123", "1234")
        test("123", 0, sequences)
        test("1234", 0, sequences)
    }

    @Test
    fun testEmpty() {
        val sequences = listOf("", "123")
        test("123", 0, sequences)
        test("1234", 0, sequences)
        test("", 0, sequences)
    }

    @Test(expected = IllegalArgumentException::class)
    fun testNoSequences() {
        val sequences = emptyList<String>()
        test("123", null, sequences)
        test("1234", null, sequences)
        test("", null, sequences)
    }

    @Test
    fun testTwoBranchesLong() {
        val sequences = listOf("123abcdef", "124abcdef")
        test("123abcdef", 0, sequences)
        test("124abcdef", 1, sequences)
        test("123abcdefXYZ", 0, sequences)
        test("124abcdefXYZ", 1, sequences)
        test("123abcde", null, sequences)
        test("124abcde", null, sequences)
        test("123abcdeX", null, sequences)
        test("124abcdeX", null, sequences)
        test("", null, sequences)
    }

    @Test
    fun testDepth() {
        testDepthInternal()
        testDepthInternal("")
        testDepthInternal("", "1")
        testDepthInternal("1", "")
        testDepthInternal("1")
        testDepthInternal("1", "2")
        testDepthInternal("1", "22")
        testDepthInternal("11", "2")
        testDepthInternal("11", "22")
        testDepthInternal("11", "222")
        testDepthInternal("111", "222")
        testDepthInternal("111", "222", "1111111")
    }

    private fun testDepthInternal(vararg sequences: String) {
        val sequencesList = sequences.toList()

        val node = sequencesMap(sequencesList).byteNode()

        val expectedDepth = sequencesList
            .map(String::length)
            .maxOrNull() ?: 0

        Assert.assertEquals(expectedDepth, node.depth)
    }

    @Test
    fun testAllBytesValues() {
        val branchId = 1
        for (branchByte in allBytes()) {
            val sequence = intArrayOf(branchByte.toInt() and 0xff)

            val sequencesTree = mapOf(sequence to branchId).byteNode()
            for (inputByte in allBytes()) {
                val inputData = byteArrayOf(inputByte)
                ByteSequenceRecognizer(
                    ByteArrayInputStream(inputData),
                    sequencesTree
                ).use { bsr ->
                    val actualBytes = bsr.stream.readAllBytes()
                    Assert.assertArrayEquals(inputData, actualBytes)
                    val actualResult = bsr.result
                    val expectedResult = if (branchByte == inputByte)
                        branchId
                    else
                        null
                    Assert.assertEquals(expectedResult, actualResult)
                }
            }
        }
    }

    private fun test(inputStr: String, expectedResult: Int?, sequences: List<String>) {
        recognizer(
            inputStr,
            sequencesMap(sequences).byteNode()
        ).use { bsr ->
            val actualText = bsr.stream.readAllBytes().string
            Assert.assertEquals(inputStr, actualText)
            val actualResult = bsr.result
            Assert.assertEquals(expectedResult, actualResult)
        }
    }

    companion object {

        private fun allBytes(): Iterable<Byte> {
            return Iterable {
                object : Iterator<Byte> {
                    var next: Byte? = Byte.MIN_VALUE

                    override fun hasNext(): Boolean {
                        return next != null
                    }

                    override fun next(): Byte {
                        val ret = next
                        next = if (next == Byte.MAX_VALUE) {
                            null
                        } else {
                            (next!! + 1).toByte()
                        }
                        return ret!!
                    }
                }
            }
        }
    }
}

private val String.intArray
    get() = this.bytes.map(Byte::toInt).toIntArray()

private fun recognizer(s: String, sequencesTree: ByteNode<Int>) =
    ByteSequenceRecognizer(s.bytes.inputStream(), sequencesTree)

private val String.bytes
    get() = this.toByteArray(TEST_CHARSET)

private val ByteArray.string
    get() = String(this, TEST_CHARSET)

private fun sequencesMap(sequences: List<String>): Map<IntArray, Int> {
    val r: MutableMap<IntArray, Int> = HashMap()
    for (i in sequences.indices) {
        r[sequences[i].intArray] = i
    }
    return r
}